src/analysis/retail/monk/mistweaver/normalizers/CastLinkNormalizer.ts

```diff
@@ -0,0 +1,173 @@
+  RemoveBuffEvent,
+} from 'parser/core/Events';
+
+export const APPLIED_HEAL = 'AppliedHeal';
+export const BOUNCED = 'Bounced';
+export const FROM_DANCING_MISTS = 'FromDM';
+export const FROM_HARDCAST = 'FromHardcast';
+export const FROM_MISTY_PEAKS = 'FromMistyPeaks';
+export const FROM_RAPID_DIFFUSION = 'FromRD'; // can be linked to env mist or rsk cast
+
+const CAST_BUFFER_MS = 150;
+const FOUND_REMS = new Set();
+
+/*
+  This file is for attributing Renewing Mist and Enveloping Mist applications to hard casts.
+  It is needed because mistweaver talents can proc ReM/EnvM, 
+  but they are not extended by RM nor do they trigger the flat RM Heal
+  */
+const EVENT_LINKS: EventLink[] = [
+  // link renewing mist apply to its CastEvent
+  {
+    linkRelation: FROM_HARDCAST,
+    reverseLinkRelation: APPLIED_HEAL,
+    linkingEventId: [SPELLS.RENEWING_MIST_HEAL.id],
+    linkingEventType: [EventType.ApplyBuff],
+    referencedEventId: TALENTS_MONK.RENEWING_MIST_TALENT.id,
+    referencedEventType: [EventType.Cast],
+    forwardBufferMs: CAST_BUFFER_MS,
+    backwardBufferMs: CAST_BUFFER_MS,
+  },
+  // link Enveloping Mist apply to its cast
+  {
+    linkRelation: FROM_HARDCAST,
+    reverseLinkRelation: APPLIED_HEAL,
+    linkingEventId: [TALENTS_MONK.ENVELOPING_MIST_TALENT.id, SPELLS.ENVELOPING_MIST_TFT.id],
+    linkingEventType: [EventType.ApplyBuff, EventType.RefreshBuff],
+    referencedEventId: TALENTS_MONK.ENVELOPING_MIST_TALENT.id,
+    referencedEventType: EventType.Cast,
+    forwardBufferMs: CAST_BUFFER_MS,
+    backwardBufferMs: CAST_BUFFER_MS,
+  },
+  // link renewing mist apply to the target it was removed from
+  {
+    linkRelation: BOUNCED,
+    linkingEventId: [SPELLS.RENEWING_MIST_HEAL.id],
+    linkingEventType: [EventType.ApplyBuff],
+    referencedEventId: SPELLS.RENEWING_MIST_HEAL.id,
+    referencedEventType: [EventType.RemoveBuff],
+    forwardBufferMs: 0,
+    backwardBufferMs: 500,
+    anyTarget: true,
+    additionalCondition(linkingEvent, referencedEvent) {
+      return (
+        (linkingEvent as ApplyBuffEvent).targetID !== (referencedEvent as RemoveBuffEvent).targetID
+      );
+    },
+  },
+  // link renewing mist removal to its application event
+  {
+    linkRelation: BOUNCED,
+    linkingEventId: [SPELLS.RENEWING_MIST_HEAL.id],
+    linkingEventType: [EventType.RemoveBuff],
+    referencedEventId: SPELLS.RENEWING_MIST_HEAL.id,
+    referencedEventType: [EventType.ApplyBuff],
+    forwardBufferMs: 0,
+    backwardBufferMs: 65000,
+  },
+  // link ReM to an EnvM/RSK cast
+  {
+    linkRelation: FROM_RAPID_DIFFUSION,
+    linkingEventId: [SPELLS.RENEWING_MIST_HEAL.id],
+    linkingEventType: [EventType.ApplyBuff],
+    referencedEventId: [
+      TALENTS_MONK.RISING_SUN_KICK_TALENT.id,
+      TALENTS_MONK.ENVELOPING_MIST_TALENT.id,
+      SPELLS.ENVELOPING_MIST_TFT.id,
+    ],
+    referencedEventType: [EventType.Cast],
+    forwardBufferMs: 500,
+    backwardBufferMs: 500,
+    anyTarget: true,
+  },
```

Comment:
Big possibility of false positive here - I'm pretty sure if player casts Enveloping Mists -> Renewing Mist, both spells will go off at the same time, and then this link will catch the hardcast.

While it's not documented, the eventlinks are applied in order so I think you could add an additional condition to this link to check that it doesn't already have a FROM_HARDCAST link. You'd be breaking new ground for this normalizer, but it was a use case I had in mind.

Reply:
Addressing this one by adding an additionalCondition that it cannot be linked to a hardcast as suggested in DMs, which should address false positive issue